Rework C++ examples

IgorAnohin opened the pull request:
Hello!
It would be interesting hear your opinion about this PR.

Commit:
C++: Navio: add libraries from Navio+

Navio2: only Navio2 libraries
Navio+: onle Navio+ libraries
Common: common libraries

Commit:
C++: Navio: Common: add Led interface

Make new interface

Commit:
C++: Navio: Common: add RCInput interface

Make new interface

Commit:
C++: Examples: ADC: add Navio+ support

Makefile: add Navio+ libraries
ADC.cpp: add Navio+ entry

Commit:
C++: Examples: Barometer: add new path to libraries

MakeFile: rename paths
Barometer: add new paths to libraries

Commit:
C++: Examples: GPS: change paths to libraries

rename from makefile to Makefile
Makefile: change paths libraries
gps: change path to libraries

Commit:
C++: Examples: LED: add Navio+ support

rename Led2
LED: rewrite file with Led interface and add error message when use without sudo
Makefile: add Navio+ libaries

Commit:
C++: Examples: add example for FRAM on Navio+

FRAM only for Navio+

Commit:
C++: Examples: RCInput: add Navio+ support

main: rewrite with RCInput interface
Makefile: add new libraries from Navio+

Commit:
C++: Examples: add Multithread Examples

Commit:
C++: Examples: Servo: add Navio+ support

Makefile: add Navio+ libraries
Servo: rewrite file with RCOutput interface and add error message when use without sudo

Commit:
C++: Navio: Common: Util: add new function

Util: add get_navio_version function

staroselskii commented on C++/Navio/Navio+/RCInput_Navio.h:
```diff
@@ -0,0 +1,50 @@
+    void initialize();
+    int read(int ch);
+    ~RCInput_Navio();
+    void ppmOnEdge(int gpio, int level, uint32_t tick);
```
Why it's public? Should be private. You can make this function a friend.

staroselskii commented on C++/Navio/Common/RCOutput.h:
```diff
@@ -0,0 +1,17 @@
+#ifndef RCOUTPUT
+#define RCOUTPUT
+
+#define PWN_OUTPUT 0
```
typo
s/PWN_OUTPUT/PWM_OUTPUT

And also it should be here

staroselskii replied:
Please, move it where it belongs. Maybe Examples? But most definitely not library code

staroselskii commented on C++/Navio/Navio2/RCOutput_Navio2.cpp:
```diff
@@ -0,0 +1,26 @@
+bool RCOutput_Navio2::initialize()
+{
+    if (!pwm.init(PWN_OUTPUT)) {
+        fprintf(stderr, "Output Enable not set. Are you root?\n");
```
Doesn't make sense. We don't have Output enable on Navio2.

IgorAnohin replied:
Sorry, but I can't agree with you. Navio2 have Output enable

staroselskii commented on C++/Navio/Navio2/RCOutput_Navio2.cpp:
```diff
@@ -0,0 +1,26 @@
+
+void RCOutput_Navio2::setFrequency(float frequency)
+{
+    pwm.set_period(PWN_OUTPUT, frequency);
```
Why only on zeroth channel? It doesn't make sense! Please.

IgorAnohin replied:
Yes, you are right, I will correct all mistakes.

staroselskii commented on C++/Navio/Common/RCOutput.h:
```diff
@@ -0,0 +1,17 @@
+{
+public:
+    virtual bool initialize() = 0;
+    virtual void setFrequency(float frequency) = 0;
```
setFrequency should work on a channel basis.

And also stick to underscores in naming.

It should be rather set_frequency.

staroselskii commented on C++/Examples/AccelGyroMag/AccelGyroMag.cpp:
```diff
@@ -47,36 +47,50 @@ void print_help()
+            return "-1";
+        }
+
+        // prevent the error message
```
What's that?

staroselskii commented on C++/Examples/AccelGyroMag/AccelGyroMag.cpp:
```diff
@@ -47,36 +47,50 @@ void print_help()
+            case 'h': print_help(); return "-1";
+            case '?': printf("Wrong parameter.\n");
+                      print_help();
+                      return "-1";
```
-1 is extremely bad! Please, say at least "none" or just use plain C strings and return a nullptr.

staroselskii commented on C++/Examples/ADC/ADC.cpp:
```diff
@@ -1,26 +1,55 @@
-            printf("A%d: %.4fV ", i, results[i] / 1000);
+    if (get_navio_version() == NAVIO2) {
+
+        ADC adc{};
```
Rename that to ADC_Navio2, add a new ADC interface and subtype. This is example code and should not deal with implementation details.

staroselskii commented on C++/Examples/ADC/ADC.cpp:
```diff
@@ -1,26 +1,55 @@
+        }
+    } else {
+
+        ADS1115 adc;
```
Put all of this in ADC_Navio

staroselskii commented on C++/Examples/LED/LED.cpp:
```diff
@@ -0,0 +1,61 @@
+        return EXIT_FAILURE;
+
+    while (true) {
+        led -> setColor(Colors::Green);
```
unneeded spaces between led and setColor

staroselskii commented on C++/Examples/Servo/Servo.cpp:
```diff
@@ -14,35 +14,51 @@ sudo ./Servo
+        pwm->setFrequency(50);
+
+        while (true) {
+            pwm->set_duty_cycle(PWN_OUTPUT, SERVO_MIN); //To change the channel, you need to change PWN_OUTPUT
```
PWN_OUTPUT

staroselskii reviewed (changes requested):
Overall a big improvement. Thanks for your hard work. Please, fix these nitpickings and I'll happily merge it.

Commit:
C++: Navio: Common: add ADC interface

Make new interface

Commit:
C++: Examples: AccelGyroMag: add Navio+ support

Makefile: add Navio+ libraries
AccelGyroMag: add auto shield choice

Commit:
C++: Examples: AHRS: add Navio+ support

Makefile: add Navio+ libraries
AHRS.cpp: add auto shield choice
AHRS.hpp: move realization functions to AHRS.cpp

IgorAnohin commented:
What are you think about this PR now?

staroselskii commented on C++/Navio/Navio2/RCOutput_Navio2.cpp:
```diff
@@ -0,0 +1,21 @@
+bool RCOutput_Navio2::initialize(int channel)
+{
+
+    pwm.enable(channel);
```
missing return

staroselskii commented on C++/Navio/Navio2/RCOutput_Navio2.cpp:
```diff
@@ -0,0 +1,21 @@
+
+bool RCOutput_Navio2::set_duty_cycle(int channel, float period)
+{
+    pwm.set_duty_cycle(channel, period / 1000);
```
ditto

staroselskii commented on C++/Navio/Navio2/RCOutput_Navio2.h:
```diff
@@ -0,0 +1,20 @@
+{
+public:
+    RCOutput_Navio2();
+    bool initialize(int channel);
```
missing `override` specifier on all of the overridden methods

staroselskii commented on C++/Navio/Navio+/RCOutput_Navio.cpp:
```diff
@@ -0,0 +1,35 @@
+        pin.write(0); /* drive Output Enable low */
+    } else {
+        fprintf(stderr, "Output Enable not set. Are you root?\n");
+        return 0;
```
return false!

staroselskii commented on C++/Navio/Navio+/RCOutput_Navio.cpp:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    pwm.initialize();
+    return 1;
```
return true or maybe return pwm.inititalize() ?

staroselskii commented on C++/Navio/Navio+/RCOutput_Navio.cpp:
```diff
@@ -0,0 +1,35 @@
+bool RCOutput_Navio::set_duty_cycle(int channel, float period)
+{
+    pwm.setPWMmS(channel+3, period / 1000); // 1st Navio RC output is 3 and period must be less 10
+
```
where's the ruturn?

staroselskii commented on C++/Navio/Navio+/RCOutput_Navio.h:
```diff
@@ -0,0 +1,23 @@
+{
+public:
+    RCOutput_Navio();
+    bool initialize(int channel);
```
`override` missing everywhere.

staroselskii commented on C++/Navio/Navio+/ADC_Navio.cpp:
```diff
@@ -0,0 +1,27 @@
+
+int ADC_Navio::read(int ch)
+{
+    adc.setMultiplexer(muxes[ch]);
```
read(100) will blow up stack. Please add a check `if ch < ARRAY_SIZE(muxes)`.

staroselskii reviewed (changes requested):
Nice but please verify that all methods that should have return values, do actually have them. Add parameter checking if need be. Like if we channel.read(100), it should just return an error.
Next step would be to turn all Navio/Navio2 and Common to a library libnavio and then just link examples to it like this gcc -o adc adc.cpp -lnavio

Commit:
C++: Navio: Navio+: add Led Navio+ support

add Led_Navio driver

Commit:
C++: Navio: Navio2: add Led Navio2 support

add Led_Navio2 driver

Commit:
C++: Navio: Naivio+: add RCInput Navio+ support

Add RCInpit_Navio driver

Commit:
C++: Navio: Navio2: add RCInput Navio2 support

Add RCInput_Navio2 driver

Commit:
C++: Navio: Navio2: add RCOutput Navio2 support

Add RCOutput_Navio2 driver

Commit:
C++: Navio: Navio2: add ADC Navio2 support

add ADC_Navio2 driver

Commit:
C++: Navio: Navio+: add ADC Navio+ support

Add ADC_Navio diver

Commit:
C++: Navio: Common: add RCOutput interface

Make new interface

Commit:
C++: Navio: Navio+: add RCOutput Navio+ support

Add RCOutput_Navio driver

IgorAnohin commented:
Hi!
I fixed bugs and add new library.

staroselskii commented on C++/Navio/Navio2/Led_Navio2.cpp:
```diff
@@ -0,0 +1,18 @@
+
+bool Led_Navio2::initialize()
+{
+    return (!pwm.initialize());
```
We don't use PWM for Leds on Navio2.
It's also extremely confusing to negate the initialize()'s return value here. Why?

staroselskii reviewed (changes requested):
Everything's fine except LEDs examples don't work on either of boards.

Commit:
C++: Navio: add library libnavio

Add new Makefile, which build navio library
Add 'Wno-psabi' flag (https://github.com/nlohmann/json/issues/658)
Add gitignore file

Commit:
C++: Examples: update Makefiles

Link examples with libnavio

staroselskii commented on C++/Navio/Navio2/RGBled.cpp:
```diff
@@ -14,7 +14,7 @@ bool RGBled::initialize() {
     pinG = new Pin(RPI_GPIO_27);
     pinB = new Pin(RPI_GPIO_6);
-    if (pinR->init() & pinG->init() & pinB->init()) {
+    if (pinR->init() && pinG->init() && pinB->init()) {
```
The change is correct but the commit message is not good.
`C++: Navio2: RGBLed: fail on logic AND in pin initialization`

The body should say that there had been bitwise and in the past.

staroselskii commented on C++/Navio/Navio2/RGBled.cpp:
```diff
@@ -25,7 +25,7 @@ bool RGBled::initialize() {
     }
     else {
-        fprintf(stderr, "Output enable not set. Are you root?\n");
+        fprintf(stderr, "Initialization failed\n");
```
It's better to give more context to an error. Like `%s: pins' initialization failed, __FUNCTION__`

Commit:
C++: Navio: Common: fix initialize bug

Change PERI_BASE for raspberry 3

Commit:
C++: Navio: Navio2: RGBled: fail on logic AND in pin initialization

Add get_failed_pin function and change & to &&
